packages/blocks/src/page-block/edgeless/mode-controllers/default-mode.ts

```diff
@@ -20,89 +21,84 @@ import {
+      return {
+        start: new DOMPoint(this._dragStartPos.x, this._dragStartPos.y),
+        end: new DOMPoint(this._dragLastPos.x, this._dragLastPos.y),
+      };
+    }
+    return null;
+  }
+
   private _pick(x: number, y: number) {
     const { surface } = this._edgeless;
     const [modelX, modelY] = surface.viewport.toModelCoord(x, y);
     const selectedShape = surface.pickTop(modelX, modelY);
-    return selectedShape ? selectedShape : pick(this._blocks, modelX, modelY);
-  }
-
-  private _updateDraggingArea(x: number, y: number) {
-    if (this._draggingArea) {
-      this._draggingArea.end = new DOMPoint(x, y);
-    }
-    if (this._hoverState) {
-      this._setSingleSelectionState(this._hoverState.content, false);
-    }
-  }
-
-  private _updateHoverState(content: Selectable | null) {
-    if (!content) {
-      this._hoverState = null;
-      return;
-    }
-
-    const { viewport } = this._edgeless.surface;
-    const xywh = getXYWH(content);
-    this._hoverState = {
-      rect: getSelectionBoxBound(viewport, xywh),
-      content,
-    };
+    return selectedShape
+      ? selectedShape
+      : pickTopBlock(this._blocks, modelX, modelY);
   }
 
   private _setNoneSelectionState() {
-    this._blockSelectionState = { type: 'none' };
-    this._edgeless.slots.updateSelection.emit(this._blockSelectionState);
+    this._blockSelectionState = { selected: [], active: false };
+    this._edgeless.slots.selectionUpdated.emit(this._blockSelectionState);
     resetNativeSelection(null);
   }
 
-  private _setSingleSelectionState(selected: Selectable, active: boolean) {
-    const { viewport } = this._edgeless.surface;
-    const xywh = getXYWH(selected);
+  private _setSelectionState(selected: Selectable[], active: boolean) {
     this._blockSelectionState = {
-      type: 'single',
-      active,
       selected,
-      rect: getSelectionBoxBound(viewport, xywh),
+      active,
     };
-    this._edgeless.slots.updateSelection.emit(this._blockSelectionState);
+    this._edgeless.slots.selectionUpdated.emit(this._blockSelectionState);
   }
 
   private _handleClickOnSelected(selected: Selectable, e: SelectionEvent) {
     const isSurfaceEl = isSurfaceElement(selected);
 
     // shape
     if (isSurfaceEl) {
-      this._setSingleSelectionState(selected, true);
+      this._setSelectionState([selected], false);
     }
     // block
     else {
-      switch (this.blockSelectionState.type) {
-        case 'none':
-          this._setSingleSelectionState(selected, false);
+      switch (this.blockSelectionState.selected.length) {
+        case 0:
+          this._setSelectionState([selected], false);
           break;
-        case 'single':
-          if (this.blockSelectionState.selected === selected) {
-            this._setSingleSelectionState(selected, true);
+        case 1:
+          if (this.blockSelectionState.selected[0] === selected) {
+            this._setSelectionState([selected], true);
```

Comment:
It's so ugly. It should be optimized.

Reply:
I think it's a bug not handling `length > 1` here.

Reply:
optimized in commit [15753a8](https://github.com/toeverything/blocksuite/pull/1533/commits/15753a8d75ad4ab99064d100f9ca803166ea53d7)